src/Features/Core/Portable/CodeRefactorings/SyncNamespace/AbstractSyncNamespaceService.MoveFileCodeAction.cs

```diff
@@ -0,0 +1,189 @@
+
+                return ImmutableArray.Create<CodeActionOperation>(
+                    new ApplyChangesOperation(solution),
+                    new OpenDocumentOperation(newDocumentId, activateIfAlreadyOpen: true));
+            }
+            
+            public static ImmutableArray<MoveFileCodeAction> Create(State state)
+            {
+                Debug.Assert(state.RelativeDeclaredNamespace != null);
+
+                // Since all documents have identical folder structure, we can do the computation on any of them.
+                var document = state.Solution.GetDocument(state.OriginalDocumentId);
+                // In case the relative namespace is "", the file should be moved to project root,
+                // set `parts` to empty to indicate that.
+                var parts = state.RelativeDeclaredNamespace.Length == 0
+                    ? ImmutableArray<string>.Empty
+                    : state.RelativeDeclaredNamespace.Split(new[] { '.' }).ToImmutableArray();
+
+                // Invalid char can only appear in namespace name when there's error,
+                // which we have checked before creating any code actions.
+                Debug.Assert(parts.IsEmpty || parts.Any(s => s.IndexOfAny(Path.GetInvalidPathChars()) < 0));
+
+                var projectRootFolder = FolderInfo.CreateFolderHierarchyForProject(document.Project);
+                var candidateFolders = FindCandidateFolders(projectRootFolder, parts, ImmutableArray<string>.Empty);
+                return candidateFolders.SelectAsArray(folders => new MoveFileCodeAction(state, folders));
+            }
+
+            /// <summary>
+            /// We try to provide additional "move file" options if we can find existing folders that matches target namespace.
+            /// For example, if the target namespace is 'DefaultNamesapce.A.B.C', and there's a folder 'ProjectRoot\A.B\' already 
+            /// exists, then will provide two actions, "move file to ProjectRoot\A.B\C\" and "move file to ProjectRoot\A\B\C\".
+            /// </summary>
+            private static ImmutableArray<ImmutableArray<string>> FindCandidateFolders(
+                FolderInfo currentFolderInfo, 
+                ImmutableArray<string> parts, 
+                ImmutableArray<string> currentFolder)
+            {
+                if (parts.IsEmpty)
+                {
+                    return ImmutableArray.Create(currentFolder);
+                }
+
+                // Try to figure out all possible folder names that can match the target namespace.
+                // For example, if the target is "A.B.C", then the matching folder names include
+                // "A", "A.B" and "A.B.C". The item "index" in the result tuple is the number 
+                // of items in namespace parts used to construct iten "foldername".
+                var candidates = Enumerable.Range(1, parts.Length)
```

Comment:
won't it be cheaper just getting list without "\" or "." and find sub or whole matching all names and get what it is pointing to? rather than recursively going in to sub folder?